File: Core/Models/Branch/BranchFinancials.swift
import Foundation

public struct BranchFinancials: Codable, Identifiable, Hashable, Sendable {
    public let id: EntityID
    public var branchID: EntityID
    public var month: Date    // first day of the calendar month
    public var revenueAED: Double
    public var rentAED: Double
    public var utilitiesAED: Double
    public var staffCostAED: Double
    public var equipmentAED: Double
    public var marketingAED: Double
    public var otherExpensesAED: Double
    public var outstandingFeesAED: Double
    public var activePaymentPlans: Int

    public init(
        id: EntityID = UUID(),
        branchID: EntityID,
        month: Date,
        revenueAED: Double,
        rentAED: Double,
        utilitiesAED: Double,
        staffCostAED: Double,
        equipmentAED: Double,
        marketingAED: Double,
        otherExpensesAED: Double,
        outstandingFeesAED: Double = 0,
        activePaymentPlans: Int = 0
    ) {
        self.id = id
        self.branchID = branchID
        self.month = month
        self.revenueAED = revenueAED
        self.rentAED = rentAED
        self.utilitiesAED = utilitiesAED
        self.staffCostAED = staffCostAED
        self.equipmentAED = equipmentAED
        self.marketingAED = marketingAED
        self.otherExpensesAED = otherExpensesAED
        self.outstandingFeesAED = outstandingFeesAED
        self.activePaymentPlans = activePaymentPlans
    }

    public var totalExpensesAED: Double {
        rentAED + utilitiesAED + staffCostAED + equipmentAED
        + marketingAED + otherExpensesAED
    }
    public var netContributionAED: Double { revenueAED - totalExpensesAED }
}

File: flutter_app/lib/core/models/branch_financials.dart
import 'entity_id.dart';

/// Port of Core/Models/Branch/BranchFinancials.swift.

class BranchFinancials {
  final EntityID id;
  final EntityID branchId;

  /// First day of the calendar month this record covers.
  final DateTime month;
  final double revenueAed;
  final double rentAed;
  final double utilitiesAed;
  final double staffCostAed;
  final double equipmentAed;
  final double marketingAed;
  final double otherExpensesAed;
  final double outstandingFeesAed;
  final int activePaymentPlans;

  const BranchFinancials({
    required this.id,
    required this.branchId,
    required this.month,
    required this.revenueAed,
    required this.rentAed,
    required this.utilitiesAed,
    required this.staffCostAed,
    required this.equipmentAed,
    required this.marketingAed,
    required this.otherExpensesAed,
    this.outstandingFeesAed = 0,
    this.activePaymentPlans = 0,
  });

  double get totalExpensesAed =>
      rentAed +
      utilitiesAed +
      staffCostAed +
      equipmentAed +
      marketingAed +
      otherExpensesAed;

  double get netContributionAed => revenueAed - totalExpensesAed;

  factory BranchFinancials.fromJson(Map<String, dynamic> json) =>
      BranchFinancials(
        id: json['id'] as String,
        branchId: json['branchID'] as String,
        month: DateTime.parse(json['month'] as String),
        revenueAed: (json['revenueAED'] as num).toDouble(),
        rentAed: (json['rentAED'] as num).toDouble(),
        utilitiesAed: (json['utilitiesAED'] as num).toDouble(),
        staffCostAed: (json['staffCostAED'] as num).toDouble(),
        equipmentAed: (json['equipmentAED'] as num).toDouble(),
        marketingAed: (json['marketingAED'] as num).toDouble(),
        otherExpensesAed: (json['otherExpensesAED'] as num).toDouble(),
        outstandingFeesAed:
            (json['outstandingFeesAED'] as num?)?.toDouble() ?? 0,
        activePaymentPlans: json['activePaymentPlans'] as int? ?? 0,
      );

  Map<String, dynamic> toJson() => {
    'id': id,
    'branchID': branchId,
    'month': month.toIso8601String(),
    'revenueAED': revenueAed,
    'rentAED': rentAed,
    'utilitiesAED': utilitiesAed,
    'staffCostAED': staffCostAed,
    'equipmentAED': equipmentAed,
    'marketingAED': marketingAed,
    'otherExpensesAED': otherExpensesAed,
    'outstandingFeesAED': outstandingFeesAed,
    'activePaymentPlans': activePaymentPlans,
  };
}
